### PR TITLE
Quick full screen

### PR DESCRIPTION
**Pull request by maxdeviant**
This PR is for investigating/adding the quick full screen functionality that @jordwalke requested in #666 (I guess I'm doing the devil's work here 😱 ).

Currently I have it working quickly by just expanding the window to take up all available space on the screen. It feels a little hacky to me, but it is fast :smile:

#### TODO:
- [x] Enter quick full screen mode
- [x] Leave quick full screen mode
  - [x] Restore the window to the previous size/location it was in before entering full screen
- [x] Toggle the menu item appropriately
- [x] Add accelerators for Linux
- [x] ~~Figure out if there is a more elegant way to do this~~
  - I tried the simple stuff first (i.e., using [`BrowserWindow.setFullScreen`](http://electron.atom.io/docs/api/browser-window/#winsetfullscreenflag), but that still uses the OS X animations. 
  - **EDIT:** After making some tweaks and sleeping on it, I think that my current implementation is probably the best way to go about it, given the constraints of Electron.

### CODE REVIEW

**Comment by jordwalke** [comment]
I hope that other electron apps can learn from your approach.

File: app/menu.js [hunk 1]
```diff
@@ -312,6 +312,15 @@ module.exports = function createMenu({createWindow, updatePlugins}) {
       },
       {
         role: 'togglefullscreen'
+      },
+      {
+        label: 'Enter Quick Full Screen',
+        accelerator: 'Cmd+Enter',
```
Comment (jordwalke) [inline app/menu.js]: Nice choice! (Is it configurable?)

File: app/index.js [hunk 2]
```diff
@@ -245,6 +246,16 @@ app.on('ready', () => installDevExtensions(isDev).then(() => {
       win.maximize();
     });
 
+    rpc.on('enter quick full screen', () => {
+      const {width, height} = electron.screen.getPrimaryDisplay().workAreaSize;
+      win.setSize(width, height);
```
Comment (jordwalke) [inline app/index.js]: Does this render on top of the menu bar?

Comment (maxdeviant) [inline app/index.js]: Not currently. I was going to ask what you had in mind for the behavior.

**Comment by maxdeviant** [comment]
@jordwalke If it works well I can see about abstracting it into an Electron plugin so other apps can add the same functionality easily 👍

**Comment by maxdeviant** [comment]
Here it is in action: https://gfycat.com/BitterScentedBarnowl

File: app/menu.js [hunk 3]
```diff
@@ -312,6 +313,28 @@ module.exports = function createMenu({createWindow, updatePlugins}) {
...
+        click(item, focusedWindow) {
+          if (focusedWindow) {
+            focusedWindow.rpc.emit('enter quick full screen req');
+          }
+        }
+      },
+      {
+        id: 'LEAVE_QUICK_FULL_SCREEN',
+        label: 'Leave Quick Full Screen',
+        accelerator: 'Cmd+Enter',
```
Comment (flybayer) [inline app/menu.js]: Please add an accelerator for Linux users :grinning: We don't have `Cmd`

Comment (maxdeviant) [inline app/menu.js]: Got it!

Comment (maxdeviant) [inline app/menu.js]: @flybayer Any thoughts on what the accelerator should be? :thinking: `Alt+Enter`?

Comment (flybayer) [inline app/menu.js]: That may be ok, although `Ctrl+Shift+Enter` might be better. It's similar to `Cmd+Shift+F` for OSX.

**Comment by rauchg** [comment]
How about ^⌘F ?

**Comment by rauchg** [comment]
(not sure it's better, but seems more standard?)

**Comment by maxdeviant** [comment]
@rauchg ^⌘F is already reserved for normal full screen on OSX (which triggers the animation).

**Comment by shime** [comment]
Greta work, can't wait for it to be merged. Worth noting that the hotkey for entering full screen in iTerm is also `⌘↵`.

**Comment by shime** [comment]
Detected a couple of issues after using it for longer time:

* it remembers the window size after exiting, so when it's restarted it will show in full screen - maybe not a problem, but it's unable to exit full screen as a result
* the event is not firing every time the hotkey is pressed, it fires after 2-3 attempts occasionally

**Comment by maxdeviant** [comment]
@shime Yea, I was aware of the first issue, but I'm not totally sure what the best way to handle it is.

Because Hyper saves your terminal position and size when it exits (as part of another feature).

I was debating between:

- Simply not saving the position/size when in quick full screen
  - This would restore the position/size you had before entering full screen on the next load
- Saving the "is full screen" state to the Electron config (where the position and size are recorded)
  - This would bring you back to full screen on the next load, but you would be able to toggle it off using the hotkey

The former is easier since we just have to track the state in-memory, as opposed to the latter where it must be written out when the app exits.

cc @rauchg @jordwalke Any input on how this should work?

**Comment by maxdeviant** [comment]
@shime Also, that second issue is interesting. I've only done cursory testing, so the next time I have a look at it I will see if I can reproduce and fix the issue.

**Comment by jordwalke** [comment]
Me? I'm not picky about which of those two solutions you go for as long as it's not buggy. I'm just happy to have full screen!

File: lib/index.js [hunk 4]
```diff
@@ -80,6 +80,14 @@ rpc.on('split request vertical', () => {
...
 });
 
+rpc.on('enter quick full screen req', () => {
+  store_.dispatch(uiActions.enterQuickFullScreen());
+});
+
+rpc.on('leave quick full screen req', () => {
+  store_.dispatch(uiActions.leaveQuickFullScreen());
+});
+
```
Comment (rauchg) [inline lib/index.js]: I like this "roundtrip" approach. Gives us lots of flexibility

**Comment by jordwalke** [comment]
Did I miss why this was reverted?

**Comment by maxdeviant** [comment]
@jordwalke There was an issue with it where sometimes you would get locked in quick full screen and not be able to get out. Usually it happened if you switched back and forth too fast.

I've been meaning to have a look at it, but haven't had any time to devote to it lately :slightly_frowning_face:

**Comment by hojberg** [comment]
damn i was excited about this

**Comment by ChadTaljaardt** [comment]
Bump for this, it is a must have feature for productivity in my opinion.

**Comment by asantos00** [comment]
Any plans about the `Cmd+enter` action to fullscreen on mac?

**Comment by albinekb** [comment]
you can do that with the custom keymaps @asantos00

**Comment by asantos00** [comment]
So it will not be added as default behaviour? I think it is default behaviour for all OSX apps :(

**Pull request merged**